vendor/bat-native-ledger/src/bat/ledger/internal/database/database_recurring_tip.cc

```diff
@@ -66,18 +69,139 @@ void DatabaseRecurringTip::InsertOrUpdate(
   ledger_->RunDBTransaction(std::move(transaction), transaction_callback);
 }
 
+void DatabaseRecurringTip::InsertOrUpdate(
+    const std::string& publisher_id,
+    double amount,
+    base::OnceCallback<void(bool)> callback) {
+  if (publisher_id.empty()) {
+    BLOG(1, "Publisher ID is empty");
+    std::move(callback).Run(false);
+    return;
+  }
+
+  if (amount <= 0) {
+    BLOG(1, "Invalid contribution amount");
+    std::move(callback).Run(false);
+    return;
+  }
+
+  auto transaction = mojom::DBTransaction::New();
+
+  const std::string query = base::StringPrintf(
+      "INSERT OR REPLACE INTO %s "
+      "(publisher_id, amount, added_date, next_contribution_at) "
+      "VALUES (?, ?, ?, ?)",
+      kTableName);
+
+  auto command = mojom::DBCommand::New();
+  command->type = mojom::DBCommand::Type::RUN;
+  command->command = query;
+
+  uint64_t added_at = util::GetCurrentTimeStamp();
+
+  BindString(command.get(), 0, publisher_id);
+  BindDouble(command.get(), 1, amount);
+  BindInt64(command.get(), 2, added_at);
+  BindInt64(command.get(), 3, added_at + constant::kReconcileInterval);
+
+  transaction->commands.push_back(std::move(command));
+
+  auto on_completed = [](base::OnceCallback<void(bool)> callback,
+                         mojom::DBCommandResponsePtr response) {
+    std::move(callback).Run(response &&
+                            response->status ==
+                                mojom::DBCommandResponse::Status::RESPONSE_OK);
+  };
+
+  ledger_->RunDBTransaction(std::move(transaction),
+                            base::BindOnce(on_completed, std::move(callback)));
+}
+
+void DatabaseRecurringTip::AdvanceMonthlyContributionDates(
+    const std::vector<std::string>& publisher_ids,
+    base::OnceCallback<void(bool)> callback) {
+  const std::string query = base::StringPrintf(
+      "UPDATE %s SET next_contribution_at = ? WHERE publisher_id = ?",
+      kTableName);
+
+  uint64_t next = util::GetCurrentTimeStamp() + constant::kReconcileInterval;
+
+  auto transaction = mojom::DBTransaction::New();
+  for (const std::string& publisher_id : publisher_ids) {
+    if (!publisher_id.empty()) {
+      auto command = mojom::DBCommand::New();
+      command->type = mojom::DBCommand::Type::RUN;
+      command->command = query;
+      BindInt64(command.get(), 0, next);
+      BindString(command.get(), 1, publisher_id);
+      transaction->commands.push_back(std::move(command));
+    }
+  }
+
+  if (transaction->commands.empty()) {
+    std::move(callback).Run(true);
+    return;
+  }
+
+  auto on_completed = [](base::OnceCallback<void(bool)> callback,
```

Comment:
I think we could break this out into its own function.

Reply:
Done.